File: JavaScript/boj/Silver/15666.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const [N, M] = input[0].split(' ').map(Number);
const numbers = input[1]
  .split(' ')
  .map(Number)
  .sort((a, b) => a - b);
const result = [];
const sequence = [];

const dfs = (depth, start) => {
  if (depth === M) {
    result.push(sequence.join(' '));
    return;
  }

  for (let i = start; i < N; i++) {
    sequence.push(numbers[i]);
    dfs(depth + 1, i);
    sequence.pop();
  }
};

dfs(0, 0);
console.log([...new Set(result)].join('\n'));

File: Python/boj/Silver/15666.py
import sys
from itertools import combinations_with_replacement

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N, M = map(int, input().split())
    numbers = list(map(int, input().split()))
    result = []
    visited = set()
    numbers.sort()

    permutation = combinations_with_replacement(numbers, M)

    for i in permutation:
        if not i in visited:
            visited.add(i)
            result.append(i)

    for j in result:
        print(" ".join(map(str, j)))
